src/AspNet.Security.OAuth.Alipay/AlipayAuthenticationHandler.cs

```diff
@@ -130,40 +144,158 @@ protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull] OA
             return new AuthenticationTicket(context.Principal, context.Properties, Scheme.Name);
         }
 
-        protected override string FormatScope() => string.Join(",", Options.Scope);
-        private IDictionary<string, string> RSA2Sign(SortedDictionary<string, string> source)
+        /// <summary>
+        /// get payload as an asynchronous operation.
+        /// </summary>
+        /// <param name="message">The message.</param>
+        /// <param name="nodeName">Name of the node.</param>
+        /// <returns>Task&lt;JsonDocument&gt;.</returns>
+        private async Task<JsonDocument> getPayloadAsync(HttpResponseMessage message, string nodeName)
         {
-            var textparams = source.Where(p => !string.IsNullOrEmpty(p.Value)).Select(p => p.Key + "=" + p.Value);
-            if (!textparams.Any())
-                throw new HttpRequestException("sign fail");
-            var rsa2 = new RSAHelper(RSAType.RSA2, Encoding.UTF8, Options.ClientSecret, Options.AlipayPublicKey);
-            source.Add("sign", rsa2.Sign(string.Join("&", textparams)));
-            return source;
+            var messageContent = await message.Content.ReadAsStringAsync();
+
+            //Logger.LogInformation(messageContent);
+
+            var rootJson = JsonDocument.Parse(messageContent);
+            var payload = JsonDocument.Parse(rootJson.RootElement.GetString(nodeName));
+
+            return payload;
+        }
+
+        /// <summary>
+        /// Gets the RSA2 signature.
+        /// </summary>
+        /// <param name="source">The source.</param>
+        /// <param name="privateKey">The private key.</param>
+        /// <returns>System.String.</returns>
+        private string getRSA2Signature(SortedDictionary<string, string> source, [NotNull] string privateKey)
+        {
+            var textParams = source
+                .Where(p => !string.Equals(p.Key, "sign", StringComparison.OrdinalIgnoreCase)
+                    && !string.IsNullOrEmpty(p.Value))
+                .Select(p => p.Key + "=" + p.Value);
+
+            var privateKeyRsaProvider = CreateRsaProviderFromPrivateKey(privateKey);
+
+            var dataBytes = Encoding.UTF8.GetBytes(string.Join("&", textParams));
+            var signatureBytes = privateKeyRsaProvider.SignData(dataBytes, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);
+
+            return Convert.ToBase64String(signatureBytes);
+        }
+
+        /// <summary>
+        /// Creates the RSA provider from private key.
+        /// </summary>
+        /// <param name="privateKey">The private key.</param>
+        /// <returns>RSA.</returns>
+        /// <exception cref="Exception">
+        /// Unexpected value read binr.ReadUInt16()
+        /// or
+        /// Unexpected version
+        /// or
+        /// Unexpected value read binr.ReadByte()
+        /// </exception>
+        private RSA CreateRsaProviderFromPrivateKey(string privateKey)
```

Comment:
Thanks - this is much clearer to its intent compared to the previous code.

As pointed out to me by @vcsjones, there's APIs in .NET Core 3.0 that allow you to do this in a much simpler way:

```csharp
private RSA CreateAlgorithm()
{
    byte[] privateKeyBytes = Convert.FromBase64String(Options.ClientSecret);
    var algorithm = RSA.Create();

    try
    {
        algorithm.ImportRSAPrivateKey(privateKeyBytes, out int _);
        return algorithm;
    }
    catch (Exception)
    {
        algorithm.Dispose();
        throw;
    }
}
```

You can see a similar usage in our Sign in with Apple provider:

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/1a32506eedbacd83fed18b985aa11d58928192d7/src/AspNet.Security.OAuth.Apple/Internal/DefaultAppleClientSecretGenerator.cs#L94-L108